release.py

```diff
@@ -0,0 +1,178 @@
+    major = int(match.group(1))
+    minor = int(match.group(2))
+    patch = int(match.group(3))
+
+    if bump_type == BumpType.MAJOR:
+        return f"{major+1}.0.0"
+    if bump_type == BumpType.MINOR:
+        return f"{major}.{minor+1}.0"
+    if bump_type == BumpType.PATCH:
+        return f"{major}.{minor}.{patch+1}"
+
+    raise ValueError(f"Expected valid bump_type, got {bump_type} instead.")
+
+
+def get_version_match(text: str, key: str) -> Optional[Match]:  # type: ignore[type-arg]
+    pattern = get_version_pattern(key)
+    return re.search(pattern, text)
+
+
+def get_bumped_text_with_bumped_version(
+    bump_type: BumpType, text: str, key: str
+) -> str:
+    """Bump the version in SemVer syntax using the provided `bump_type`.
+
+    Usage:
+        get_bumped_version(BumpType.MAJOR, "awesome_version = 0.2.2", key = "awesome_version)
+
+        returns
+
+        "awesome_version = 1.0.0"
+    """
+    pattern = get_version_pattern(key)
+    replaced = re.sub(
+        pattern, lambda match: f'{key} = "{bump(match, bump_type)}"', text
+    )
+    return replaced
+
+
+def bump_version_in_file(
+    filename: Path, key: str, bump_type: BumpType = BumpType.PATCH
+) -> None:
+    with open(filename, "r+") as file:
+        content = file.read()
+        new_content = get_bumped_text_with_bumped_version(bump_type, content, key)
+        file.seek(0)
+        file.write(new_content)
+
+
+def get_version_from_file(filename: Path) -> Optional[Match]:  # type: ignore[type-arg]
+    with open(filename) as file:
+        return get_version_match(file.read(), KEY_MAPPING[filename])
+
+
+def _bump_version(bump_type: BumpType) -> None:
+    version = get_version_from_file(PYPROJECT_TOML)
+
+    if not version:
+        raise ValueError(f"Could not read version from file {PYPROJECT_TOML}.")
+
+    current_version = ".".join(version.groups())
+
+    next_version = bump(version, bump_type)
+
+    should_bump = typer.confirm(
+        f"Upgrade version from {current_version} to {next_version}?"
+    )
+    if not should_bump:
+        raise typer.Abort()
+
+    bump_version_in_file(PYPROJECT_TOML, key="version", bump_type=bump_type)
+    bump_version_in_file(INIT_PY, key="__version__", bump_type=bump_type)
+
+
+def commit_and_tag() -> None:
+    match = get_version_from_file(PYPROJECT_TOML)
+    if not match:
+        raise ValueError(f"Could not read version from file {PYPROJECT_TOML}.")
+
+    new_version = ".".join(match.groups())
+
+    commands = [
+        ("git", "add", str(PYPROJECT_TOML), str(INIT_PY), str(CHANGELOG)),
+        ("git", "commit", "-m", f"chore: release version {new_version}"),
+        ("git", "tag", "-s", f"{new_version}", "-m", f"Release version {new_version}"),
+    ]
+
+    for command in commands:
+        typer.echo(" ".join(command))
+        result = subprocess.call(command)
+        if result != 0:
+            raise typer.Exit(result)
```

Comment:
Setting pipefail (`set -o pipefail` I think) would be the best practice in bash. Is it possible with typer ?

Reply:
Hmm, I don't think it's a `typer` thing, we are not really doing any bash scripting here, we are just executing shell commands.